plugins/regrass.cpp

```diff
@@ -12,109 +27,553 @@ using namespace DFHack;
+    { struct_field_info::PRIMITIVE, "ashes",        offsetof(regrass_options, ashes),        &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "buildings",    offsetof(regrass_options, buildings),    &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "mud",          offsetof(regrass_options, mud),          &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "block",        offsetof(regrass_options, block),        &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "zlevel",       offsetof(regrass_options, zlevel),       &df::identity_traits<bool>::identity, 0, 0 },
+    { struct_field_info::PRIMITIVE, "forced_plant", offsetof(regrass_options, forced_plant), &df::identity_traits<int32_t>::identity, 0, 0 },
+    { struct_field_info::END }
+};
+struct_identity regrass_options::_identity(sizeof(regrass_options), &df::allocator_fn<regrass_options>, NULL, "regrass_options", NULL, regrass_options_fields);
+
+command_result df_regrass(color_ostream& out, vector<string> &parameters);
+
+static bool valid_tile(regrass_options options, df::map_block *block, int x, int y)
+{   // Is valid tile for regrass
+    auto des = block->designation[x][y];
+    auto tt = block->tiletype[x][y];
+    auto shape = tileShape(tt);
+    auto mat = tileMaterial(tt);
+    auto spec = tileSpecial(tt);
+
+    if (mat == tiletype_material::GRASS_LIGHT ||
+        mat == tiletype_material::GRASS_DARK ||
+        mat == tiletype_material::PLANT) // Shrubs and saplings can have grass underneath
+    {   // Refill existing grass
+        DEBUG(log).print("Valid tile: Grass/Shrub/Sapling\n");
+        return true;
+    }
+    else if (tt == tiletype::TreeTrunkPillar)
+    {   // Trees can have grass under main tile
+        auto p = df::coord(block->map_pos.x + x, block->map_pos.y + y, block->map_pos.z);
+        auto mbc = world->map.column_index[(block->map_pos.x / 48)*3][(block->map_pos.y / 48)*3];
+        if (!mbc)
+        {
+            Core::printerr("No MLT block column for tile at (%d, %d, %d)!\n", p.x, p.y, p.z);
+            return false;
+        }
+
+        for (auto plant : mbc->plants)
+        {
+            if (plant->pos == p)
+            {
+                DEBUG(log).print("Valid tile: Tree\n");
+                return true; // Is main tile
+            }
+        }
+
+        DEBUG(log).print("Invalid tile: Tree\n");
+        return false; // Not main tile
+    }
+    else if (des.bits.flow_size > (des.bits.liquid_type == tile_liquid::Magma ? 0 : 3))
+    {   // Under water/magma (df::plant_raw::shrub_drown_level is usually 4)
+        DEBUG(log).print("Invalid tile: Liquid\n");
+        return false;
+    }
+    else if (shape != tiletype_shape::FLOOR &&
+        shape != tiletype_shape::RAMP &&
+        shape != tiletype_shape::STAIR_UP &&
+        shape != tiletype_shape::STAIR_DOWN &&
+        shape != tiletype_shape::STAIR_UPDOWN)
```

Comment:
Saplings and shrubs are handled by `tiletype_material::PLANT` at the top.

Decided to ignore boulders and pebbles for now because they're unique features. You can smooth and build over them first if you want to turn them into grass. I could make it an option. (I'll probably end up reverse engineering at some point to verify how DF actually handles all these various cases, and make that the default.)

Reply:
got it. SGTM

Reply:
Actually, looks like some boulders on my embark have grass under them? I'll do that, then.